src/Playroom/CodeEditor/keymaps/comment.ts

```diff
@@ -0,0 +1,325 @@
+  const isInlineComment = cm
+    .getLine(from.line)
+    .trimStart()
+    .startsWith(LINE_COMMENT_START);
+  const isBlockComment =
+    cm.getLine(from.line).trimStart().startsWith(BLOCK_COMMENT_START) &&
+    cm.getLine(to.line).trimEnd().endsWith(BLOCK_COMMENT_END);
+
+  if (isInlineComment) {
+    return 'line';
+  }
+
+  if (isJavaScriptMode && !isBlockComment && !containsTag) {
+    return 'line';
+  }
+
+  if (
+    (!isJavaScriptMode && !containsAttribute) ||
+    containsTag ||
+    isJavaScriptMode
+  ) {
+    return 'block';
+  }
+
+  return 'line';
+};
+
+export const toggleComment = (cm: Editor) => {
+  const newSelections: Selection[] = [];
+  const tagRanges: TagRange[] = [];
+
+  for (const range of cm.listSelections()) {
+    const from = range.from();
+    let to = range.to();
+
+    if (to.line !== from.line && to.ch === 0) {
+      to = new Pos(to.line - 1);
+    }
+
+    const commentType = determineCommentType(cm, from, to);
+
+    const fullContent = cm.getRange(new Pos(from.line, 0), new Pos(to.line));
+    const existingIndent = fullContent.length - fullContent.trimStart().length;
+
+    const trimmedContent = fullContent.trim();
+
+    const isAlreadyCommented =
+      (trimmedContent.startsWith(BLOCK_COMMENT_START) &&
+        trimmedContent.endsWith(BLOCK_COMMENT_END)) ||
+      trimmedContent.startsWith(LINE_COMMENT_START);
+
+    const isMultiLineSelection = to.line !== from.line;
+
+    tagRanges.push({
+      from,
+      to,
+      multiLine: isMultiLineSelection,
+      existingIndent,
+      isAlreadyCommented,
+      commentType,
+    });
+
+    const fromOffset = getSelectionFromOffset({
+      commentType,
+      isAlreadyCommented,
+      fullContent,
+      from,
+    });
+
+    const toOffset = getSelectionToOffset({
+      to,
+      commentType,
+      isAlreadyCommented,
+      isMultiLineSelection,
+      fullContent,
+    });
+
+    const newSelectionRangeFrom = new Pos(from.line, from.ch + fromOffset);
+    const newSelectionRangeTo = new Pos(to.line, to.ch + toOffset);
+
+    const newSelection = isReverseSelection(range)
+      ? { anchor: newSelectionRangeTo, head: newSelectionRangeFrom }
+      : { anchor: newSelectionRangeFrom, head: newSelectionRangeTo };
+
+    newSelections.push(newSelection);
+  }
+
+  cm.operation(() => {
+    for (const range of [...tagRanges].reverse()) {
+      const newRangeFrom = new Pos(range.from.line, range.existingIndent);
+      const newRangeTo = new Pos(range.to.line);
+
+      const existingContent = cm.getRange(newRangeFrom, newRangeTo);
+
+      if (range.isAlreadyCommented) {
+        const uncommentType: CommentType = existingContent
+          .trimStart()
+          .startsWith(BLOCK_COMMENT_START)
+          ? 'block'
+          : 'line';
+
+        const existingContentWithoutComment = existingContent.replace(
+          uncommentType === 'block'
+            ? /\{\/\*\s?|\s?\*\/\}/g
+            : /^(\s*)\/\/\s?/gm,
```

Comment:
Not super concerned about perf TBH given the testing we've done, but it's still worth pulling any regexes out of this function into top-level consts, both so they don't get created each time the operation is run, and also to give them names so it's obvious what they're for.